sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java

```diff
@@ -75,191 +83,230 @@ public static Row nullRow(RowType rowType) {
+  }
+
+  /**
+   * Get a {@link TypeName#STRING} value by field name, {@link IllegalStateException} is thrown
+   * if schema doesn't match.
    */
   public String getString(String fieldName) {
-    return getValue(fieldName);
+    return getString(getSchema().indexOf(fieldName));
   }
 
   /**
-   * Get a {@link Date} value by field name, {@link ClassCastException} is thrown
-   * if type doesn't match.
+   * Get a {@link TypeName#DATETIME} value by field name, {@link IllegalStateException} is thrown
+   * if schema doesn't match.
    */
-  public Date getDate(String fieldName) {
-    return getValue(fieldName);
+  public ReadableDateTime getDateTime(String fieldName) {
+    return getDateTime(getSchema().indexOf(fieldName));
   }
 
   /**
-   * Get a {@link GregorianCalendar} value by field name, {@link ClassCastException} is thrown
-   * if type doesn't match.
+   * Get a {@link TypeName#BOOLEAN} value by field name, {@link IllegalStateException} is thrown
+   * if schema doesn't match.
    */
-  public GregorianCalendar getGregorianCalendar(String fieldName) {
-    return getValue(fieldName);
+  public boolean getBoolean(String fieldName) {
+    return getBoolean(getSchema().indexOf(fieldName));
   }
 
   /**
-   * Get a {@link BigDecimal} value by field name, {@link ClassCastException} is thrown
-   * if type doesn't match.
+   * Get an array value by field name, {@link IllegalStateException} is thrown
+   * if schema doesn't match.
    */
-  public BigDecimal getBigDecimal(String fieldName) {
-    return getValue(fieldName);
+  public <T> List<T> getArray(String fieldName) {
+    return getArray(getSchema().indexOf(fieldName));
   }
 
   /**
-   * Get a {@link Boolean} value by field name, {@link ClassCastException} is thrown
-   * if type doesn't match.
+   * Get a {@link TypeName#ROW} value by field name, {@link IllegalStateException} is thrown
+   * if schema doesn't match.
    */
-  public Boolean getBoolean(String fieldName) {
-    return getValue(fieldName);
+  public Row getRow(String fieldName) {
+    return getRow(getSchema().indexOf(fieldName));
   }
 
   /**
-   * Get a {@link Byte} value by field index, {@link ClassCastException} is thrown
-   * if type doesn't match.
+   * Get a {@link TypeName#BYTE} value by field index, {@link ClassCastException} is thrown
+   * if schema doesn't match.
    */
   public Byte getByte(int idx) {
+    checkState(
+        TypeName.BYTE.equals(getSchema().getField(idx).getTypeDescriptor().getType()));
     return getValue(idx);
   }
 
   /**
-   * Get a {@link Short} value by field index, {@link ClassCastException} is thrown
-   * if type doesn't match.
+   * Get a {@link TypeName#INT16} value by field index, {@link ClassCastException} is thrown
+   * if schema doesn't match.
    */
-  public Short getShort(int idx) {
+  public Short getInt16(int idx) {
+    checkState(
+        TypeName.INT16.equals(getSchema().getField(idx).getTypeDescriptor().getType()));
```

Comment:
I think that ClassCastExecptions better describe the issue here than IllegalStateExceptions